include/pico_config.h

```diff
@@ -229,7 +235,43 @@ static inline uint64_t long_long_be(uint64_t le)
 # include "arch/pico_posix.h"
 #endif
 
-#ifdef PICO_SUPPORT_MM
+
+#ifdef CHECK_MEM
+extern int start_failing_mallocs;
+
+static inline void log_malloc(const char* filename, const char* message)
+{
+	FILE * file;
+	file = fopen(filename, "a");
+	fprintf(file, message);
+	fclose(file);
+}
+
+static inline void append_backtrace(const char* filename)
+{
+	void *array[10];
+	size_t size;
+	int fd;
+	FILE * file;
+	file = fopen(filename, "a");
+	fd = fileno(file);
+
+	fprintf(file, "Backtrace:\n");
+	fseek(file, 0, SEEK_END);
+
+	size = backtrace(array, 10);
+	backtrace_symbols_fd(array, size, fd);
+	fprintf(file, "\n");
+
+	fclose(file);
+}
+
+#define PICO_ZALLOC(x) \
+((start_failing_mallocs && (((double)(rand())/(double)RAND_MAX) < 0.4)) \
+? (log_malloc("mem_test.log", "Malloc FAILED\n"), append_backtrace("mem_test.log"),  NULL) \
+: (log_malloc("mem_test.log", "Malloc Succeeded\n"), append_backtrace("mem_test.log"), pico_zalloc(x))) 
```

Comment:
An afterthought : it looks like this statement expression is GNU C [0], not entirely standard. In the interest of compiler compatibility, it may be better to just put this into a function...

[0] https://gcc.gnu.org/onlinedocs/gcc-3.3.1/gcc/Statement-Exprs.html